tests/unit-tests/test-class-usage-tracking.php

```diff
@@ -0,0 +1,318 @@
+			return function( $message, $title, $args ) {
+				$e = new WP_Die_Exception( 'wp_die called' );
+				$e->set_wp_die_args( $message, $title, $args );
+				throw $e;
+			};
+		} );
+	}
+
+	/* END test ajax request cases */
+
+	/**
+	 * Ensure that a request is made to the correct URL with the given
+	 * properties and the default properties.
+	 */
+	public function test_send_event() {
+		$event      = 'my_event';
+		$properties = array(
+			'button_clicked' => 'my_button'
+		);
+		$timestamp  = '1234';
+
+		// Capture the network request, save the request URL and arguments, and
+		// simulate a WP_Error
+		$request_params = null;
+		$request_url    = null;
+		add_filter( 'pre_http_request', function( $preempt, $r, $url ) use ( &$request_params, &$request_url ) {
+			$request_params = $r;
+			$request_url    = $url;
+			return new WP_Error();
+		}, 10, 3 );
+
+		Sensei_Usage_Tracking::send_event( 'my_event', $properties, $timestamp );
+
+		$parsed_url = parse_url( $request_url );
+
+		$this->assertEquals( 'pixel.wp.com', $parsed_url['host'] );
+		$this->assertEquals( '/t.gif', $parsed_url['path'] );
+
+		$query = array();
+		parse_str( $parsed_url['query'], $query );
+		$this->assertArraySubset( array(
+			'button_clicked' => 'my_button',
+			'admin_email'    => 'admin@example.org',
+			'_ut'            => 'sensei:site_url',
+			'_ui'            => 'http://example.org',
+			'_ul'            => '',
+			'_en'            => 'sensei_my_event',
+			'_ts'            => '1234000',
+			'_'              => '_',
+		), $query );
+	}
+
+	/**
+	 * Ensure that the request is only sent when the setting is enabled.
+	 */
+	public function test_maybe_send_usage_data() {
+		$count = 0;
+
+		// Count the number of network requests
+		add_filter( 'pre_http_request', function() use ( &$count ) {
+			$count++;
+			return new WP_Error();
+		} );
+
+		// Setting is not set, ensure the request is not sent.
+		Sensei_Usage_Tracking::maybe_send_usage_data();
+		$this->assertEquals( 0, $count );
+
+		// Set the setting and ensure request is sent.
+		Sensei()->settings->set( 'sensei_usage_tracking_enabled', true );
+		Sensei()->settings->get_settings();
+
+		Sensei_Usage_Tracking::maybe_send_usage_data();
+		$this->assertEquals( 1, $count );
+	}
+
+	/* Tests for tracking opt in dialog */
+
+	/**
+	 * When setting is not set, dialog is not hidden, and user has capability,
+	 * we should see the dialog and Enable Usage Tracking button.
+	 */
+	public function test_display_tracking_opt_in() {
+		$this->setup_opt_in_dialog();
+
+		$this->expectOutputRegex( '/Enable Usage Tracking/' );
+		$this->usage_tracking->maybe_display_tracking_opt_in();
+	}
+
+	/**
+	 * When setting is already set, dialog should not appear.
+	 */
+	public function test_do_not_display_tracking_opt_in_when_setting_enabled() {
+		$this->setup_opt_in_dialog();
+		Sensei()->settings->set( 'sensei_usage_tracking_enabled', true );
+		Sensei()->settings->get_settings();
+
+		$this->expectOutputString( '' );
```

Comment:
I'm not quite following these last few tests and why `expectOutputString` is being called before the call to `maybe_display_tracking_opt_in`. Could you clarify?

Reply:
It's because of the way [output testing](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.output) works in phpunit. You need to set up the expectation first (probably because it starts the buffering at this point) and then call the code that produces the output.

Reply:
Oh, interesting! TIL.